compiler/rustc_parse/src/parser/diagnostics.rs

```diff
@@ -1076,6 +1120,145 @@ impl<'a> Parser<'a> {
+        };
+
+        self.recover_from_inc_dec(operand_expr, kind, op_span)
+    }
+
+    pub(super) fn maybe_recover_from_postfix_increment(
+        &mut self,
+        operand_expr: P<Expr>,
+        op_span: Span,
+        prev_is_semi: bool,
+    ) -> PResult<'a, P<Expr>> {
+        let kind = IncDecRecovery {
+            standalone: prev_is_semi,
+            op: IncOrDec::Inc,
+            fixity: UnaryFixity::Post,
+        };
+
+        self.recover_from_inc_dec(operand_expr, kind, op_span)
+    }
+
+    fn recover_from_inc_dec(
+        &mut self,
+        base: P<Expr>,
+        kind: IncDecRecovery,
+        op_span: Span,
+    ) -> PResult<'a, P<Expr>> {
+        let mut err = self.struct_span_err(
+            op_span,
+            &format!("Rust has no {} {} operator", kind.fixity, kind.op.name()),
+        );
+        err.span_label(op_span, &format!("not a valid {} operator", kind.fixity));
+
+        if let ExprKind::Path(_, path) = &base.kind {
+            if let [segment] = path.segments.as_slice() {
+                let ident = segment.ident;
+                // (pre, post)
+                let spans = match kind.fixity {
+                    UnaryFixity::Pre => (op_span, ident.span.shrink_to_hi()),
+                    UnaryFixity::Post => (ident.span.shrink_to_lo(), op_span),
+                };
+
+                if !ident.is_reserved() {
+                    if kind.standalone {
+                        return self.inc_dec_standalone_recovery(base, err, kind, ident, spans);
+                    } else {
+                        match kind.fixity {
+                            UnaryFixity::Pre => {
+                                return self.prefix_inc_dec_suggest_and_recover(
+                                    base, err, kind, ident, spans,
+                                );
+                            }
+                            UnaryFixity::Post => {
+                                return self.postfix_inc_dec_suggest_and_recover(
+                                    base, err, kind, ident, spans,
+                                );
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // base case
+        err.help(&format!("use `{}= 1` instead", kind.op.chr()));
+        err.emit();
+
+        Ok(base)
+    }
+
+    fn prefix_inc_dec_suggest_and_recover(
+        &mut self,
+        base: P<Expr>,
+        mut err: DiagnosticBuilder<'_>,
+        kind: IncDecRecovery,
+        ident: Ident,
+        (pre_span, post_span): (Span, Span),
+    ) -> PResult<'a, P<Expr>> {
+        err.multipart_suggestion(
+            &format!("use `{}= 1` instead", kind.op.chr()),
+            vec![
+                (pre_span, "{ ".to_string()),
+                (post_span, format!(" {}= 1; {} }}", kind.op.chr(), ident)),
+            ],
+            Applicability::MachineApplicable,
+        );
+        err.emit();
+        // TODO: recover
```

Comment:
It seems like recovering from this might be fairly involved. Is it worth the added code complexity to return an expression like `{ i += 1; i }` instead of having the user just fix their code and rebuild to see further errors?

Reply:
At least initially, if recovery is challenging to implement, we can skip that so that this can land.

Reply:
Sounds good.